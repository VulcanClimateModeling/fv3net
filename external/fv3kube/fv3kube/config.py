import os
import datetime
from pathlib import Path
from typing import Any, Sequence, Mapping, Optional
import fsspec
import yaml
import fv3config


# Map for different base fv3config dictionaries
PWD = Path(os.path.abspath(__file__)).parent
DEFAULT_BASE_VERSION = "v0.3"  # for backward compatibility
BASE_FV3CONFIG_BY_VERSION = {
    "v0.2": os.path.join(PWD, "base_yamls/v0.2/fv3config.yml"),
    "v0.3": os.path.join(PWD, "base_yamls/v0.3/fv3config.yml"),
    "v0.4": os.path.join(PWD, "base_yamls/v0.4/fv3config.yml"),
    "v0.5": os.path.join(PWD, "base_yamls/v0.5/fv3config.yml"),
    "v0.6": os.path.join(PWD, "base_yamls/v0.6/fv3config.yml"),
}

TILE_COORDS_FILENAMES = range(1, 7)  # tile numbering in model output filenames
RESTART_CATEGORIES = ["fv_core.res", "sfc_data", "fv_tracer.res", "fv_srf_wnd.res"]
FV_CORE_ASSET = fv3config.get_asset_dict(
    "gs://vcm-fv3config/data/initial_conditions/fv_core_79_levels/v1.0/",
    "fv_core.res.nc",
    target_location="INPUT",
)
FV3Config = Mapping[str, Any]


def _merge_once(source, update):
    """Recursively update a mapping with new values.

    Args:
        source: Mapping to be updated.
        update: Mapping whose key-value pairs will update those in source.
            Key-value pairs will be inserted for keys in update that do not exist
            in source.

    Returns:
        Recursively updated mapping.
    """
    for key in update:
        if key in ["patch_files", "diagnostics"]:
            source.setdefault(key, []).extend(update[key])
        elif (
            key in source
            and isinstance(source[key], Mapping)
            and isinstance(update[key], Mapping)
        ):
            _merge_once(source[key], update[key])
        else:
            source[key] = update[key]
    return source


def merge_fv3config_overlays(*mappings) -> Mapping:
    """Recursive merge dictionaries updating from left to right.

    For example, the rightmost mapping will override the proceeding ones. """
    out, rest = mappings[0], mappings[1:]
    for mapping in rest:
        out = _merge_once(out, mapping)
    return out


def get_base_fv3config(version_key: Optional[str] = None) -> FV3Config:
    """
    Get base configuration dictionary labeled by version_key.
    """
    version_key = version_key or DEFAULT_BASE_VERSION
    config_path = BASE_FV3CONFIG_BY_VERSION[version_key]
    with fsspec.open(config_path) as f:
        base_yaml = yaml.safe_load(f)

    return base_yaml


def update_tiled_asset_names(
    source_url: str,
    source_filename: str,
    target_url: str,
    target_filename: str,
    **kwargs,
) -> Sequence[Mapping[str, str]]:

    """
    Update tile-based fv3config assets with new names.  Uses format to update
    any data in filename strings with category, tile, and provided keyword
    arguments.

    Filename strings should include any specified variable name inserts to
    be updated with a format. E.g., "{timestep}.{category}.tile{tile}.nc"
    """
    assets = [
        fv3config.get_asset_dict(
            source_url,
            source_filename.format(category=category, tile=tile, **kwargs),
            target_location=target_url,
            target_name=target_filename.format(category=category, tile=tile, **kwargs),
        )
        for category in RESTART_CATEGORIES
        for tile in TILE_COORDS_FILENAMES
    ]

    return assets


def get_full_config(
    config_update: FV3Config, ic_url: str, ic_timestep: str
) -> FV3Config:
    """Given config_update return full fv3config object pointing to initial conditions
    at {ic_url}/{ic_timestep}. Initial condition filenames assumed to include prepended
    timestamp.

    Args:
        config_update: fv3config update object containing parameters different from
            the base config specified by its "base_version" argument
        ic_url: path to directory containing all initial conditions
        ic_timestep: timestamp of desired initial condition

    Returns:
        fv3config Mapping
    """
    base_version = config_update.get("base_version", DEFAULT_BASE_VERSION)
    return merge_fv3config_overlays(
        get_base_fv3config(base_version),
        c48_initial_conditions_overlay(ic_url, ic_timestep),
        config_update,
    )


def c48_initial_conditions_overlay(url: str, timestep: str) -> Mapping:
    """An overlay containing initial conditions namelist settings
    """
    TIME_FMT = "%Y%m%d.%H%M%S"
    time = datetime.datetime.strptime(timestep, TIME_FMT)
    time_list = [time.year, time.month, time.day, time.hour, time.minute, time.second]

    overlay = {}
    overlay["initial_conditions"] = update_tiled_asset_names(
        source_url=os.path.join(url, timestep),
        source_filename="{timestep}.{category}.tile{tile}.nc",
        target_url="INPUT",
        target_filename="{category}.tile{tile}.nc",
        timestep=timestep,
    )
    overlay["initial_conditions"].append(FV_CORE_ASSET)
    overlay["namelist"] = {}
    overlay["namelist"]["coupler_nml"] = {
        "current_date": time_list,
        "force_date_from_namelist": True,
    }

    return overlay
