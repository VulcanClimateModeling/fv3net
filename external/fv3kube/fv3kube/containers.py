import os
import yaml
from typing import Mapping, Optional

from kubernetes.client import (
    V1ResourceRequirements,
    V1Toleration,
    V1PodSpec,
    V1Container,
    V1VolumeMount,
    V1Volume,
    V1SecretVolumeSource,
    V1EmptyDirVolumeSource,
    V1EnvVar,
    V1ObjectMeta,
    V1JobSpec,
    V1Job,
    V1PodTemplateSpec,
)


def insert_gcp_secret(container: V1Container, secret_vol: V1Volume):
    """Insert GCP credentials into a container

    Args:
        container: a container to insert the gcp credentials into
        secret_vol: a volume containg a GCP service account key at
            ``key.json``
            
    """
    if container.volume_mounts is None:
        container.volume_mounts = []

    if container.env is None:
        container.env = []

    container.volume_mounts.append(
        V1VolumeMount(mount_path="/etc/gcp/", name=secret_vol.name)
    )

    container.env.extend(
        [
            V1EnvVar(name="GOOGLE_APPLICATION_CREDENTIALS", value="/etc/gcp/key.json"),
            V1EnvVar(
                name="CLOUDSDK_AUTH_CREDENTIAL_FILE_OVERRIDE", value="/etc/gcp/key.json"
            ),
        ]
    )


script_template = """
import fv3config
import yaml
import os

model_url = os.environ["MODEL"]
config = yaml.safe_load(os.environ["FV3CONFIG"])
fv3config.write_run_directory(config, "{rundir}")
with open("{rundir}/fv3config.yml", "w") as f:
    yaml.safe_dump(config, f)
"""


def write_rundir_container(
    config, data_vol: V1Volume, secret_vol: V1Volume, image: str
) -> V1Container:
    """Container for writing run-directories with fv3config

    Args:
        config: an fv3config dictionary
        data_vol: a k8s volume to store the data in. The data will be written
            to the location "rundir" in this volume.
        secret_vol: a volume contain a GCP service account key at the path
            ``key.json``
        image: an docker image with fv3config installed
    Returns:
        a k8s container that will write the run directory
        
    """
    container = V1Container(name="write-rundir")
    container.volume_mounts = [
        V1VolumeMount(mount_path="/mnt", name=data_vol.name),
    ]

    # Inject the fv3config and runfile via an environmental variable
    container.env = [
        V1EnvVar(name="FV3CONFIG", value=yaml.safe_dump(config)),
        V1EnvVar(name="MODEL", value=yaml.safe_dump(config)),
    ]
    container.image = image
    container.command = ["python", "-c"]
    container.args = [script_template.format(rundir="/mnt/rundir")]

    insert_gcp_secret(container, secret_vol)

    return container


fv3_run_template = """
2>&1 mpirun -np 6 \
    --oversubscribe \
    --allow-run-as-root \
    --oversubscribe \
    --allow-run-as-root\
    --mca btl_vader_single_copy_mechanism none \
    python {runfile} \
    | tee logs.txt

echo $? > fv3_exit_status
# always exit 0 so upload step proceeds
exit 0
"""


def fv3_container(data_vol: V1Volume, image: str, cpu: str, memory: str) -> V1Container:
    """A container for running a run-directory from a volume

    Args:
        data_vol: A k8s volume containing a path "rundir" with a saved run-directory.
        image: the prognostic_run docker image

    Returns:
        container for running the output. stores data in the volume at path
        "rundir". To allow future containers to run, this container will
        always exit 0, but it will save the exit status to a file
        ``fv3_exit_status``. The logs will be saved to ``logs.txt``.
        
     """
    runfile_path = "/fv3net/workflows/prognostic_c48_run/sklearn_runfile.py"
    fv3_container = V1Container(name="fv3")
    fv3_container.image = image
    fv3_container.working_dir = "/mnt/rundir"
    fv3_container.command = ["bash", "-c"]
    fv3_container.args = [fv3_run_template.format(runfile=runfile_path)]
    fv3_container.resources = V1ResourceRequirements(
        limits=dict(cpu=cpu, memory=memory), requests=dict(cpu=cpu, memory=memory),
    )

    fv3_container.volume_mounts = [
        V1VolumeMount(mount_path="/mnt", name=data_vol.name),
    ]

    return fv3_container


def post_process_container(
    path: str,
    destination: str,
    data_vol: V1Volume,
    secret_vol: V1Volume,
    image: str = "us.gcr.io/vcm-ml/post_process_run:latest",
) -> V1Container:
    """Container for post processing fv3 model output for cloud storage

    Args:
        path: relative path within volume pointing to run-directory
        destination: uri to desired GCS output directory
        vol: a K8s volume containing ``path`
        image: the docker image to use for post processing.
    Returns
        a k8s container encapsulating the post-processing.
    """
    container = V1Container(name="post")
    container.volume_mounts = [
        V1VolumeMount(mount_path="/mnt", name=data_vol.name),
    ]

    rundir = os.path.join("/mnt", path)
    container.image = image
    container.command = ["post_process.py", rundir, destination]
    # Suitable for C48 job
    container.resources = V1ResourceRequirements(
        limits=dict(cpu="6", memory="3600M"), requests=dict(cpu="6", memory="3600M"),
    )

    insert_gcp_secret(container, secret_vol)

    return container


def post_processed_fv3_pod_spec(
    model_config: Mapping,
    output_url: str,
    fv3config_image: str,
    fv3_image: str,
    post_process_image: str,
    gcp_secret_name: str = "gcp-key",
    cpu: str = "6",
    memory: str = "6Gi",
) -> V1PodSpec:
    """A PodSpec for running the prognostic run
    
    Runs FV3 with the fv3config object ``model_config`` and saves the post
    processed output to ``output_url``.

    This orchestrates three containers. The first writes the run-directory,
    the second runs the model, and third post-processes and uploads the data.
    """

    empty_vol = V1Volume(name="rundir")
    empty_vol.empty_dir = V1EmptyDirVolumeSource()

    secret_vol = V1Volume(name="google-secret")
    secret_vol.secret = V1SecretVolumeSource(secret_name=gcp_secret_name)

    # Need to add toleration for large jobs
    climate_sim_toleration = V1Toleration(
        effect="NoSchedule", value="climate-sim-pool", key="dedicated", operator="Equal"
    )

    return V1PodSpec(
        init_containers=[
            write_rundir_container(
                model_config, empty_vol, secret_vol, fv3config_image
            ),
            fv3_container(empty_vol, image=fv3_image, cpu=cpu, memory=memory),
        ],
        containers=[
            post_process_container(
                "rundir", output_url, empty_vol, secret_vol, image=post_process_image
            )
        ],
        volumes=[empty_vol, secret_vol],
        restart_policy="Never",
        tolerations=[climate_sim_toleration],
    )


def pod_spec_to_job(
    pod_spec: V1PodSpec,
    labels: Mapping[str, str],
    generate_name: Optional[str] = None,
    name: Optional[str] = None,
    backoff_limit: int = 0,
) -> V1Job:

    template_spec = V1PodTemplateSpec(
        metadata=V1ObjectMeta(labels=labels), spec=pod_spec,
    )
    job_spec = V1JobSpec(
        template=template_spec,
        backoff_limit=backoff_limit,
        completions=1,
        ttl_seconds_after_finished=100,
    )
    job = V1Job(
        api_version="batch/v1",
        kind="Job",
        metadata=V1ObjectMeta(name=name, generate_name=generate_name, labels=labels),
        spec=job_spec,
    )
    return job
