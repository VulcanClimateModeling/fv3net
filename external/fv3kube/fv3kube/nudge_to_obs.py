from datetime import datetime, timedelta
import os
import numpy as np
from typing import List, Mapping

import fsspec
import fv3config

# this module assumes that analysis files are at 00Z, 06Z, 12Z and 18Z
SECONDS_IN_HOUR = 60 * 60
NUDGE_HOURS = np.array([0, 6, 12, 18])  # hours at which analysis data is available
NUDGE_FILE_TARGET = "INPUT"  # where to put analysis files in rundir


def _most_recent_nudge_time(start_time: datetime) -> datetime:
    """Return datetime object for the last nudging time preceding or concurrent
     with start_time"""
    first_nudge_hour = _most_recent_hour(start_time.hour)
    return datetime(start_time.year, start_time.month, start_time.day, first_nudge_hour)


def _most_recent_hour(current_hour, hour_array=NUDGE_HOURS) -> int:
    """Return latest hour in hour_array that precedes or is concurrent with
    current_hour"""
    first_nudge_hour = hour_array[np.argmax(hour_array > current_hour) - 1]
    return first_nudge_hour


def _get_nudge_time_list(config: Mapping) -> List[datetime]:
    """Return list of datetime objects corresponding to times at which analysis files
    are required for nudging for a given model run configuration"""
    current_date = config["namelist"]["coupler_nml"]["current_date"]
    start_time = datetime(*current_date)
    first_nudge_time = _most_recent_nudge_time(start_time)
    run_duration = fv3config.get_run_duration(config)
    nudge_duration = run_duration + (start_time - first_nudge_time)
    nudge_duration_hours = int(
        np.ceil(nudge_duration.total_seconds() / SECONDS_IN_HOUR)
    )
    nudge_interval = NUDGE_HOURS[1] - NUDGE_HOURS[0]
    nudging_hours = range(0, nudge_duration_hours + nudge_interval, nudge_interval)
    return [first_nudge_time + timedelta(hours=hour) for hour in nudging_hours]


def _get_nudge_filename_list(config: Mapping) -> List[str]:
    """Return list of filenames of all nudging files required"""
    nudge_filename_pattern = config["gfs_analysis_data"]["filename_pattern"]
    time_list = _get_nudge_time_list(config)
    return [time.strftime(nudge_filename_pattern) for time in time_list]


def _get_nudge_files_asset_list(config: Mapping) -> List[Mapping]:
    """Return list of fv3config assets for all nudging files required for a given
    model run configuration"""
    nudge_url = config["gfs_analysis_data"]["url"]
    return [
        fv3config.get_asset_dict(nudge_url, file, target_location=NUDGE_FILE_TARGET)
        for file in _get_nudge_filename_list(config)
    ]


def _get_nudge_files_description_asset(config: Mapping, config_url: str) -> Mapping:
    """Return an fv3config asset pointing to the text file that the
    model requires to describe the list of nudging files."""
    fname_list_filename = config["namelist"]["fv_nwp_nudge_nml"]["input_fname_list"]
    return fv3config.get_asset_dict(config_url, fname_list_filename)


def _get_input_fname_list_asset(config: Mapping, filename: str) -> Mapping:
    fname_list_contents = "\n".join(_get_nudge_filename_list(config))
    data = fname_list_contents.encode()
    return fv3config.get_bytes_asset_dict(
        data, target_location="", target_name=filename
    )


def enable_nudge_to_observations(config: Mapping) -> Mapping:
    """Return an configuration dictionary with nudging to observations enabled

    Accepts and returns an fv3config dictionary

    Note:
        This function appends to patch_files and alters the namelist
    """
    input_fname_list = "nudging_file_list"
    config = _assoc_nudging_namelist_options(config, input_fname_list=input_fname_list)
    fname_list_asset = _get_input_fname_list_asset(config, input_fname_list)

    patch_files = config.setdefault("patch_files", [])
    patch_files.append(fname_list_asset)
    patch_files.append(_get_nudge_files_asset_list(config))

    return config


def _assoc_nudging_namelist_options(
    config,
    gfs_analysis_url="gs://vcm-ml-data/2019-12-02-year-2016-T85-nudging-data",
    input_fname_list="nudging_file_list",
    tau_ps=21600.0,
    tau_virt=21600.0,
    tau_winds=21600.0,
    tau_q=21600.0,
) -> Mapping:
    """assoc is a common name for adding new items to a dictionary without mutation"""

    # TODO Oli, please indicate which options below are unrelated to nudging
    namelist_overlay = {
        "gfs_analysis_data": {
            "url": gfs_analysis_url,
            "filename_pattern": "%Y%m%d_%HZ_T85LR.nc",
        },
        "namelist": {
            "atmos_model_nml": {"fhout": 2.0, "fhmax": 10000},
            "fv_core_nml": {"nudge": True},
            "gfs_physics_nml": {"fhzero": 2.0, "use_analysis_sst": True},
            "fv_nwp_nudge_nml": {
                "add_bg_wind": False,
                "do_ps_bias": False,
                "ibtrack": True,
                "input_fname_list": input_fname_list,
                "k_breed": 10,
                "kbot_winds": 0,
                "mask_fac": 0.2,
                "nf_ps": 3,
                "nf_t": 3,
                "nudge_debug": True,
                "nudge_hght": False,
                "nudge_ps": True,
                "nudge_virt": True,
                "nudge_winds": True,
                "nudge_q": True,
                "r_hi": 5.0,
                "r_lo": 3.0,
                "r_min": 225000.0,
                "t_is_tv": False,
                "tau_ps": tau_ps,
                "tau_virt": tau_virt,
                "tau_winds": tau_winds,
                "tau_q": tau_q,
                "tc_mask": True,
                "time_varying": False,
                "track_file_name": "No_File_specified",
                "use_high_top": True,
            },
        },
    }

    fv3config_overlay = {"namelist": namelist_overlay}
    return vcm.update_nested_dict(config, fv3config_overlay)
