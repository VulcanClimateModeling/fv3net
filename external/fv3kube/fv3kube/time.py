from typing import List
import datetime
import cftime

TIME_FMT = "%Y%m%d.%H%M%S"


def decode_time(timestamp: str) -> cftime.DatetimeJulian:
    return datetime.datetime.strptime(timestamp, TIME_FMT)


def encode_time(time: cftime.DatetimeJulian) -> str:
    return time.strftime(TIME_FMT)


def time_to_list(t: cftime.DatetimeJulian) -> List[int]:
    """Retrieve the 'current_date' in the format required by fv3gfs namelist
    from timestamp string."""
    return [t.year, t.month, t.day, t.hour, t.minute, t.second]