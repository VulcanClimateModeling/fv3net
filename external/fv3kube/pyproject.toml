[tool.poetry]
name = "fv3kube"
version = "0.1.0"
description = ""
authors = ["Oliver Watt-Meyer <oliwm@vulcan.com>"]

[tool.poetry.dependencies]
fsspec = "^0.6"
fv3config = { path = "../fv3config/" }
python = "^3.7"
numpy = "^1.14"
kubernetes = "^11"
pyyaml = "^5.3"


[tool.poetry.dev-dependencies]
pytest = "^5.2"

[build-system]
requires = ["poetry>=0.12"]
build-backend = "poetry.masonry.api"
