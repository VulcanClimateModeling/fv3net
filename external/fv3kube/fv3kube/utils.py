import logging
import secrets
import string
import time
import kubernetes
from kubernetes.client import BatchV1Api
from typing import Sequence, Mapping, Tuple

logger = logging.getLogger(__name__)

JobInfo = Tuple[str, str]


def initialize_batch_client() -> kubernetes.client.BatchV1Api:

    try:
        kubernetes.config.load_kube_config()
    except (TypeError, kubernetes.config.config_exception.ConfigException):
        kubernetes.config.load_incluster_config()

    batch_client = kubernetes.client.BatchV1Api()

    return batch_client


def wait_for_complete(
    job_labels: Mapping[str, str],
    batch_client: kubernetes.client.BatchV1Api = None,
    sleep_interval: int = 30,
    raise_on_fail: bool = True,
):
    """
    Function to block operation until a group of submitted kubernetes jobs complete.

    Args:
        job_labels: key-value pairs of job labels used to filter the job query. These
            values are translated to equivalence statements for the selector. E.g.,
            "key1=value1,key2=value2,..."
        client: A BatchV1Api client to communicate with the cluster.
        sleep_interval: time interval between active job check queries
        raise_on_bool: flag for whether to raise error if any job has failed

    Raises:
        ValueError if any of the jobs fail, when the
        failed jobs are first detected (if raise_on_fail is True)


    """

    if batch_client is None:
        batch_client = initialize_batch_client()

    # Check active jobs
    while True:
        time.sleep(sleep_interval)
        jobs = list_jobs(batch_client, job_labels)
        job_done = _handle_jobs(jobs, raise_on_fail)

        if job_done:
            break

    logger.info("All kubernetes jobs succesfully complete!")


def _handle_jobs(jobs: Sequence, raise_on_fail: bool) -> bool:

    failed = {}
    complete = {}
    active = {}

    for job in jobs:
        name = job.metadata.name
        if job_failed(job):
            failed[name] = job
        elif job_complete(job):
            complete[name] = job
        else:
            active[name] = job

    if len(failed) > 0:
        failed_job_names = list(failed)
        if raise_on_fail:
            raise ValueError(f"These jobs have failed: {failed_job_names}")
        else:
            logger.warning(f"These jobs have failed: {failed_job_names}")

    if len(active) == 0:
        return True
    else:
        log_active_jobs(active)
        return False


def log_active_jobs(active):
    if len(active) > 0:
        jobs = list(active.keys())
        logger.info(
            f"Active Jobs at {time.strftime('%Y-%m-%d %H:%M:%S')}\n" + "\n".join(jobs)
        )


def list_jobs(client, job_labels):
    selector_format_labels = [f"{key}={value}" for key, value in job_labels.items()]
    combined_selectors = ",".join(selector_format_labels)
    jobs = client.list_job_for_all_namespaces(label_selector=combined_selectors)

    return jobs.items


def job_failed(job):
    conds = job.status.conditions
    conditions = [] if conds is None else conds
    for cond in conditions:
        if cond.status == "True":
            return cond.type == "Failed"
    return False


def job_complete(job):
    conds = job.status.conditions
    conditions = [] if conds is None else conds
    for cond in conditions:
        if cond.status == "True":
            return cond.type == "Complete"
    return False


def delete_completed_jobs(job_labels: Mapping[str, str], client: BatchV1Api = None):
    client = initialize_batch_client() if client is None else client

    logger.info("Deleting succesful jobs.")
    jobs = list_jobs(client, job_labels)
    for job in jobs:
        if job_complete(job):
            name = job.metadata.name
            logger.info(f"Deleting completed job: {name}")
            client.delete_namespaced_job(name, namespace=job.metadata.namespace)


def get_alphanumeric_unique_tag(tag_length: int) -> str:
    """Generates a random alphanumeric string (a-z0-9) of a specified length"""

    if tag_length < 1:
        raise ValueError("Unique tag length should be 1 or greater.")

    use_chars = string.ascii_lowercase + string.digits
    short_id = "".join([secrets.choice(use_chars) for i in range(tag_length)])
    return short_id
