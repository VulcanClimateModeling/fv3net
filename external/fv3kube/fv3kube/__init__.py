from .config import (
    update_tiled_asset_names,
    get_base_fv3config,
    get_full_config,
    c48_initial_conditions_overlay,
    merge_fv3config_overlays,
)
from .utils import (
    wait_for_complete,
    delete_completed_jobs,
    job_failed,
    job_complete,
    initialize_batch_client,
    load_kube_config,
    get_alphanumeric_unique_tag,
)

__version__ = "0.1.0"
