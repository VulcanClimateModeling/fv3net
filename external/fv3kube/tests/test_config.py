import fv3kube
import pytest


@pytest.mark.parametrize(
    "mappings, expected",
    (
        [[{"a": 1}, {"a": 2}], {"a": 2}],
        [[{"a": 1}, {"b": 3}], {"a": 1, "b": 3}],
        [[{"a": 1}, {"b": 3}, {"a": 2}], {"a": 2, "b": 3}],
        [[{"a": 1}, {"a": 2, "b": 3}], {"a": 2, "b": 3}],
        [[{"a": 1}, {"a": 2}], {"a": 2}],
        [[{"a": 1}, {"a": {"b": 2}}], {"a": {"b": 2}}],
        [[{"a": {"b": 1}}, {"a": {"b": 2}}], {"a": {"b": 2}}],
        [[{"a": 1}, {"a": 1, "b": 3}], {"a": 1, "b": 3}],
        [[{"a": 1}, {"a": 2, "b": 3}], {"a": 2, "b": 3}],
        [
            [{"patch_files": ["one"]}, {"patch_files": ["two"]}],
            {"patch_files": ["one", "two"]},
        ],
        [
            [
                {"diagnostics": ["temp"]},
                {"patch_files": ["one"], "diagnostics": ["sphum"]},
            ],
            {"patch_files": ["one"], "diagnostics": ["temp", "sphum"]},
        ],
    ),
)
def test_merge_fv3config_overlays(mappings, expected):
    output = fv3kube.merge_fv3config_overlays(*mappings)
    assert output == expected


def test_invalid_vertical_coordinate_file():

    base_url = "/some/path"
    timestep = "20160805.000000"
    vertical_coordinate_file = "/some/path/"

    with pytest.raises(
        AssertionError, match="Provided vertical coordinate file is a directory"
    ):
        fv3kube.c48_initial_conditions_overlay(
            base_url, timestep, vertical_coordinate_file
        )


@pytest.mark.parametrize("vertical_coordinate_file", [None, "/some/path"])
def test_c48_initial_conditions_overlay(regtest, vertical_coordinate_file):

    url = "some/url"
    timestep = "20160801.000000"

    if vertical_coordinate_file is None:
        ans = fv3kube.c48_initial_conditions_overlay(url, timestep)
    else:
        ans = fv3kube.c48_initial_conditions_overlay(
            url, timestep, vertical_coordinate_file
        )

    print(ans, file=regtest)
