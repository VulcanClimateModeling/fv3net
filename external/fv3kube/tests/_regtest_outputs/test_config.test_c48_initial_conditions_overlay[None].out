{'initial_conditions': [{'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_core.res.tile1.nc', 'target_location': 'INPUT', 'target_name': 'fv_core.res.tile1.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_core.res.tile2.nc', 'target_location': 'INPUT', 'target_name': 'fv_core.res.tile2.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_core.res.tile3.nc', 'target_location': 'INPUT', 'target_name': 'fv_core.res.tile3.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_core.res.tile4.nc', 'target_location': 'INPUT', 'target_name': 'fv_core.res.tile4.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_core.res.tile5.nc', 'target_location': 'INPUT', 'target_name': 'fv_core.res.tile5.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_core.res.tile6.nc', 'target_location': 'INPUT', 'target_name': 'fv_core.res.tile6.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.sfc_data.tile1.nc', 'target_location': 'INPUT', 'target_name': 'sfc_data.tile1.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.sfc_data.tile2.nc', 'target_location': 'INPUT', 'target_name': 'sfc_data.tile2.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.sfc_data.tile3.nc', 'target_location': 'INPUT', 'target_name': 'sfc_data.tile3.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.sfc_data.tile4.nc', 'target_location': 'INPUT', 'target_name': 'sfc_data.tile4.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.sfc_data.tile5.nc', 'target_location': 'INPUT', 'target_name': 'sfc_data.tile5.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.sfc_data.tile6.nc', 'target_location': 'INPUT', 'target_name': 'sfc_data.tile6.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_tracer.res.tile1.nc', 'target_location': 'INPUT', 'target_name': 'fv_tracer.res.tile1.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_tracer.res.tile2.nc', 'target_location': 'INPUT', 'target_name': 'fv_tracer.res.tile2.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_tracer.res.tile3.nc', 'target_location': 'INPUT', 'target_name': 'fv_tracer.res.tile3.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_tracer.res.tile4.nc', 'target_location': 'INPUT', 'target_name': 'fv_tracer.res.tile4.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_tracer.res.tile5.nc', 'target_location': 'INPUT', 'target_name': 'fv_tracer.res.tile5.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_tracer.res.tile6.nc', 'target_location': 'INPUT', 'target_name': 'fv_tracer.res.tile6.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_srf_wnd.res.tile1.nc', 'target_location': 'INPUT', 'target_name': 'fv_srf_wnd.res.tile1.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_srf_wnd.res.tile2.nc', 'target_location': 'INPUT', 'target_name': 'fv_srf_wnd.res.tile2.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_srf_wnd.res.tile3.nc', 'target_location': 'INPUT', 'target_name': 'fv_srf_wnd.res.tile3.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_srf_wnd.res.tile4.nc', 'target_location': 'INPUT', 'target_name': 'fv_srf_wnd.res.tile4.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_srf_wnd.res.tile5.nc', 'target_location': 'INPUT', 'target_name': 'fv_srf_wnd.res.tile5.nc', 'copy_method': 'copy'}, {'source_location': 'some/url/20160801.000000', 'source_name': '20160801.000000.fv_srf_wnd.res.tile6.nc', 'target_location': 'INPUT', 'target_name': 'fv_srf_wnd.res.tile6.nc', 'copy_method': 'copy'}, {'source_location': 'gs://vcm-fv3config/data/initial_conditions/fv_core_79_levels/v1.0/', 'source_name': 'fv_core.res.nc', 'target_location': 'INPUT', 'target_name': 'fv_core.res.nc', 'copy_method': 'copy'}], 'namelist': {'coupler_nml': {'current_date': [2016, 8, 1, 0, 0, 0], 'force_date_from_namelist': True}}}
